tensorflow/python/autograph/tests/loop_scoping_test.py

```diff
@@ -38,6 +38,39 @@ def while_with_local_var(x):
     x -= 1
   return s
 
+def for_with_lambda_iter(l):
+  fns = []
+  results = []
+  for i in l:
+    fns.append(lambda: i)
+  for f in fns:
+    results.append(f())
+  return results
+
+def for_with_lambda_object():
+  class SomeRandomObject:
+    def bar(self, n):
+      return n+1
+  def foo_init():
+      return tf.constant(0)
+  fns = []
+  results = []
+  foo = foo_init()
+  for i in tf.range(3):
+    foo = SomeRandomObject()
+    fns.append( lambda i=i: foo.bar(i))
```

Comment:
I've added this test. As you can see from the CI log the output conversion it is going to fail for the live-out nature of `foo` and its unsupported type for an output conversion:

```python
TypeError: To be compatible with tf.function, Python functions must return zero or more Tensors or ExtensionTypes or None values; in compilation of <function while_loop.<locals>.wrapped_body at 0x7f0187dd6320>, found return value of type SomeRandomObject, which is not a Tensor or ExtensionType.
```

Reply:
https://github.com/tensorflow/tensorflow/blob/509c23abdf5e3ac68d588c674ba2601fef965d9e/tensorflow/python/framework/func_graph.py#L1234-L1242

Reply:
Yes, that summarizes our compatibility issue, and the reason why the fix cannot be submitted directly, without an opt-in mechanism (or some other compatibility means). So unfortunately our PR is blocked by such a compatibility mechanism.

As an aside, the error itself is one of the underlying causes of a higher level limitation: you are limited in what kind of types you can use as loop variable in a tf.while_loop. Ideally, the error message should say something in the lines of "foo is of type SomeRandomObject, which is not compatible with tf.while_loop".

Reply:
> As an aside, the error itself is one of the underlying causes of a higher level limitation: you are limited in what kind of types you can use as loop variable in a tf.while_loop. 

Can we just simply fallback lambda to local in the case we have an unsupported type?

Reply:
Possibly, but the solution would be quite complex. Basically, we'd need some flag somewhere specifying which loop var is due to a lambda capture, and just before building the tf.while_loop, discard lambda captures of unsupported types.
Anyway, my greater worry is that lambda captures may or may not become loop vars based on type, and this would be a surprising inconsistency in behavior that would be difficult to reason about.

Reply:
Yes generally I don't like to accumulate edge cases but if you see the inconsistent behavior was not strictly introduced by this PR. 

E.g. see this test without using `lamba`:

```python
@tf.function
def test(): 
  class SomeRandomObject:
    def bar(self, n):
      return n+1
  def foo_init():
      return tf.constant(0)
  foo = foo_init()
  for i in tf.range(3):
    foo = SomeRandomObject()
  return foo

test()
```

```python
    TypeError: To be compatible with tf.function, Python functions must return zero or more Tensors or ExtensionTypes or None values; in compilation of <function while_loop.<locals>.wrapped_body at 0x7f5e425bba70>, found return value of type SomeRandomObject, which is not a Tensor or ExtensionType.
```

We have exactly the same error.

I have no visibility on the no-OSS code/tests in the repo but I suppose that nobody has wrote code like this example in the internal repository (as it fails). 
So my guess is that somebody written de-facto exactly the same code as this gist but in the lambda flavor so it was not failing just cause our divergence on the variable scope between Python lambda and autograph.

So it seems that the current behavior divergence of lambda variables scope between TF/autograph and python was used to hack the limit of `tf.while_loop` return type.

> Anyway, my greater worry is that lambda captures may or may not become loop vars based on type, and this would be a surprising inconsistency in behavior that would be difficult to reason about.

Yes but if you see this failing gist it is also quite surprising that instead it works with the lambda formulation (in master).